### PR TITLE
[Sublime] Initial code for Sublime Security email connector

### PR DESCRIPTION
**Pull request by Rurik**
### Proposed changes

* This is the initial base code for a Sublime Security email external-import connector.
* Code has been tested, and formatted/verified with black and flask

### Related issues

* N/A

### Checklist

- [ ] I consider the submitted work as finished
- [ ] I have signed my commits using GPG key.
- [X] I tested the code for its functionality using different use cases
- [X] I added/update the relevant documentation (either on github or on notion)
- [X] Where necessary I refactored code to improve the overall quality

### CODE REVIEW

File: external-import/sublime/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+version: "3"
+services:
+  connector-sublime:
+    build: .
+    container_name: sublime-connector
```
Comment (helene-nguyen) [inline external-import/sublime/docker-compose.yml]: For applying the same pattern as the other connectors, here are the suggested change

```suggestion
    container_name: opencti/connector-sublime:rolling
```

The `rolling` tag is the one used to indicate that the image built is aligned with master branch version and when a release is done, a commit with release tag updates the version :)

File: external-import/sublime/src/sublime_connector.py [hunk 2]
```diff
@@ -0,0 +1,1767 @@
...
+        config_file_path = os.path.join(
+            os.path.dirname(os.path.abspath(__file__)), "config.yml"
+        )
+        config_dict = {}
+
+        if os.path.isfile(config_file_path):
+            try:
+                with open(config_file_path, "r", encoding="utf-8") as config_file:
+                    config_dict = yaml.safe_load(config_file)
+                self.helper.log_info(
```
Comment (helene-nguyen) [inline external-import/sublime/src/sublime_connector.py]: Error on running the connector as here, you use the helper before instanciating it
```Traceback (most recent call last):
File "/opt/opencti-connector-sublime/sublime_connector.py", line 44, in __init__
self.helper.log_info(
^^^^^^^^^^^
AttributeError: 'SublimeConnector' object has no attribute 'helper'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
File "/opt/opencti-connector-sublime/sublime_connector.py", line 1768, in <module>
connector = SublimeConnector()
^^^^^^^^^^^^^^^^^^
File "/opt/opencti-connector-sublime/sublime_connector.py", line 48, in __init__
self.helper.log_warning(
^^^^^^^^^^^
AttributeError: 'SublimeConnector' object has no attribute 'helper'```

File: external-import/sublime/src/sublime_connector.py [hunk 3]
```diff
@@ -0,0 +1,1767 @@
...
+            duration_obj = isodate.parse_duration(duration_period)
+            self.poll_interval = int(duration_obj.total_seconds())
+
+        except (isodate.ISO8601Error, ValueError) as e:
+            self.helper.log_warning(
+                '[!] Invalid duration format "{}": {}. Using default 5 minutes'.format(
+                    duration_period, e
+                )
+            )
+            self.poll_interval = 300  # 5 minutes default
```
Comment (helene-nguyen) [inline external-import/sublime/src/sublime_connector.py]: If duration_period is not in ISO 8601 format, an error is raised. 

Note that since the `duration_period` variable is passed directly to the helper, validation is already partially handled there. 
The code validates `duration_period` against ISO 8601 format and logs a warning on failure, defaulting to 5 minutes. 

Note: Requires traceback logging in `/__main__` to prevent silent failures.

Plus, why poll_interval is 5 minutes but duration_period (which is the time of scheduling) is 3 min ?

File: external-import/sublime/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      - CONNECTOR_DURATION_PERIOD=PT3M
+      - SUBLIME_URL=https://platform.sublime.security
+      - SUBLIME_TOKEN=ChangeMe
+      - SUBLIME_INCIDENT_TYPE=phishing  # STIX incident type
+      - SUBLIME_INCIDENT_PREFIX=Sublime Alert - 
+      - SUBLIME_CASE_PREFIX=Sublime - 
+      - SUBLIME_AUTO_CREATE_CASES=false  # Automatically create cases for incidents
+      - SUBLIME_VERDICTS=malicious  # Multiple verdicts can comma-separated like: malicious,suspicious
+      - SUBLIME_SET_PRIORITY=true  # Enable priority mapping from attack score verdict
+      - SUBLIME_SET_SEVERITY=true  # Enable severity mapping from attack score verdict
```
Comment (helene-nguyen) [inline external-import/sublime/docker-compose.yml]: Variable set_severity: 

Value differs between code and Docker. Docker is set to "True" while README is set to "False" ?

File: external-import/sublime/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      - OPENCTI_TOKEN=ChangeMe
+      - CONNECTOR_ID=a2c156d3-3bbe-4170-b370-bf6faebb56e2
+      - CONNECTOR_NAME=Sublime Security
+      - CONNECTOR_SCOPE=sublime
+      - CONNECTOR_LOG_LEVEL=info
+      - CONNECTOR_DURATION_PERIOD=PT3M
+      - SUBLIME_URL=https://platform.sublime.security
+      - SUBLIME_TOKEN=ChangeMe
+      - SUBLIME_INCIDENT_TYPE=phishing  # STIX incident type
+      - SUBLIME_INCIDENT_PREFIX=Sublime Alert -
```
Comment (helene-nguyen) [inline external-import/sublime/docker-compose.yml]: **Inconsistent default values**

Variable sublime_incident_prefix: 
Differs between code + README ("Sublime Incident - ") and Docker ("Sublime Alert - ")

File: external-import/sublime/src/sublime_connector.py [hunk 6]
```diff
@@ -0,0 +1,1767 @@
...
+
+        except Exception as e:
+            error_message = "Batch processing failed after {} messages: {}".format(
+                total_processed, e
+            )
+            self.helper.log_error("[!] {}".format(error_message))
+            self.helper.api.work.to_received(work_id, error_message)
+            raise
+
+    def run(self):
```
Comment (helene-nguyen) [inline external-import/sublime/src/sublime_connector.py:1711]: About the scheduling main process, please find in our template some best practices:

- https://github.com/OpenCTI-Platform/connectors/blob/ce8714d5d87ac6f948ab65a0d14fd6595129ca5c/templates/external-import/src/external_import_connector/connector.py#L196

And full workflow here:
- https://filigran.io/auto-backpressue-control-octi-connectors/#h-purpose-of-the-scheduler

The examples included are intended to simplify the original connector deployment and to be able to simply work when being deployed from scratch.

This method allows you to schedule the process to run at a certain intervals. This specific scheduler from the pycti connector helper will also check the queue size of a connector. 
If  `CONNECTOR_QUEUE_THRESHOLD` is set, if the connector's queue size exceeds the queue threshold, the connector's main process will not run until the queue is ingested and reduced sufficiently, allowing it to restart during the next scheduler check. (default is 500MB) It requires the duration_period connector variable in ISO-8601 standard format

File: external-import/sublime/src/sublime_connector.py [hunk 7]
```diff
@@ -0,0 +1,1767 @@
...
+            {
+                "Authorization": "Bearer {}".format(self.api_token),
+                "Accept": "application/json",
+                "User-Agent": "OpenCTI-SublimeConnector/1.0",
+            }
+        )
+
+        # Create Sublime Identity for STIX objects
+        self.sublime_identity = stix2.Identity(
+            id=self._make_deterministic_id("identity", "sublime-security-organization"),
```
Comment (helene-nguyen) [inline external-import/sublime/src/sublime_connector.py]: In the helper, you will find for entities a `generate_id` method that helps you create a deterministic ID aligned with OpenCTI platform ID resolution. This is the deduplication mechanism.

Here is the Filigran documentation:
- [Deduplication mechamism](https://docs.opencti.io/latest/usage/deduplication/)

Here an example: 
- [Link to templates README](https://github.com/OpenCTI-Platform/connectors/blob/master/templates/README.md#:~:text=To%20create%20a%20STIX,ID%2C%20use%20pycti%20library%3A)

We recommand you to use this method instead :)

**Comment by Rurik** [comment]
@helene-nguyen Acknowledged, thank you. We had some embarrassing uncaught regression errors in there. We're currentlyreviewing with our engineering team and building a new test environment to try and catch edge cases.

File: external-import/sublime/docker-compose.yml (hunk 1 above)
Comment (Rurik) [reply external-import/sublime/docker-compose.yml]: Corrected docker-compose.yml to connector-sublime and with a `:rolling` image tag

File: external-import/sublime/src/sublime_connector.py (hunk 2 above)
Comment (Rurik) [reply external-import/sublime/src/sublime_connector.py]: Good catch. This was extra troubleshooting code that ended out of sync. The log messages were removed completely in [`96240eb`](https://github.com/OpenCTI-Platform/connectors/pull/4359/commits/96240ebc204d4e4ea28116d282fe3e7f24a85a2e)

File: external-import/sublime/src/sublime_connector.py (hunk 3 above)
Comment (Rurik) [reply external-import/sublime/src/sublime_connector.py]: This block of code is removed completely (da0f4f3) I removed the interval polling and now just use the built-in OpenCTI schedular. I verified that all durations stay as ISO 8601 and are not converted to int.

File: external-import/sublime/src/sublime_connector.py (hunk 7 above)
Comment (Rurik) [reply external-import/sublime/src/sublime_connector.py]: Thank you! I don't know why or how I missed this the first time around.
In cefdb6c I removed `_make_deterministic_id()` completely and now consistently use `Incident.generate_id` or `stix2.x(value)` calls.

File: external-import/sublime/src/sublime_connector.py (hunk 6 above)
Comment (Rurik) [reply external-import/sublime/src/sublime_connector.py:1711]: Thank you. In e4fee44 I replaced my run() completely with the call to `self.helper.schedule_iso()`.

In 1976cee I corrected my use of duration_period to use the configured value and not use the one in the pycti helper, per docs.

File: external-import/sublime/docker-compose.yml (hunk 4 above)
Comment (Rurik) [reply external-import/sublime/docker-compose.yml]: I've now sync'd all values to be consistent across docker-compose.yml, config.yml.sample, and README.md

File: external-import/sublime/docker-compose.yml (hunk 5 above)
Comment (Rurik) [reply external-import/sublime/docker-compose.yml]: I've sync'd the values across all config files so that they're now consistent.

File: external-import/sublime/src/sublime_connector.py [hunk 8]
```diff
@@ -0,0 +1,1721 @@
...
+        """
+        self.helper.schedule_iso(
+            message_callback=self._process_messages,
+            duration_period=self.duration_period,
+        )
+
+
+if __name__ == "__main__":
+    connector = SublimeConnector()
+    connector.run()
```
Comment (jabesq) [inline external-import/sublime/src/sublime_connector.py]: **suggestion:** Wrap this in `try:...except Exception:` to avoid any silent failures

```suggestion
    try:
        connector = SublimeConnector()
        connector.run()
    except Exception as e:
        print(e)
        time.sleep(10)
        sys.exit(0)
```

Comment (Rurik) [reply external-import/sublime/src/sublime_connector.py]: @jabesq Thank you, and absolutely agree. Is sys.exit(0) okay, or is there an appropriate exit code for failed connectors?

Comment (jabesq) [reply external-import/sublime/src/sublime_connector.py]: Yes, sys.exit(0) is acceptable since we don't have any dedicated exit code for failed connectors. However, I agree that using sys.exit(1) would be better according to Unix conventions for exceptions which are not KeyboardInterrupt

**Review by jabesq — APPROVED** [review]

File: external-import/sublime/src/sublime_connector.py (hunk 8 above)
Comment (Rurik) [reply external-import/sublime/src/sublime_connector.py]: Thank you. sys.exit(1) fits within our internal standards, and is a bit more future proof IMO :)
